src/messages/textDocument_documentSymbol.cc

```diff
@@ -112,92 +118,100 @@ struct Handler_TextDocumentDocumentSymbol
       std::sort(out.result.begin(), out.result.end());
       pipeline::WriteStdout(kMethodType, out);
     } else if (g_config->client.hierarchicalDocumentSymbolSupport) {
-      std::unordered_map<
-          SymbolIdx, std::pair<const void *, std::unique_ptr<lsDocumentSymbol>>>
-          sym2ds;
+      std::unordered_map<SymbolIdx, std::unique_ptr<lsDocumentSymbol>> sym2ds;
+      std::vector<std::pair<const QueryFunc::Def *, lsDocumentSymbol *>> funcs;
+      std::vector<std::pair<const QueryType::Def *, lsDocumentSymbol *>> types;
       for (auto [sym, refcnt] : symbol2refcnt) {
         if (refcnt <= 0)
           continue;
         auto r = sym2ds.try_emplace(SymbolIdx{sym.usr, sym.kind});
         if (!r.second)
           continue;
-        auto &kv = r.first->second;
-        kv.second = std::make_unique<lsDocumentSymbol>();
-        lsDocumentSymbol &ds = *kv.second;
+        auto &ds = r.first->second;
+        ds = std::make_unique<lsDocumentSymbol>();
+        const void* def_ptr = nullptr;
         WithEntity(db, sym, [&](const auto &entity) {
           auto *def = entity.AnyDef();
           if (!def)
             return;
-          ds.name = def->Name(false);
-          ds.detail = def->Name(true);
+          ds->name = def->Name(false);
+          ds->detail = def->Name(true);
+
+          // Try to find a definition with spell first.
+          const void* candidate_def_ptr = nullptr;
           for (auto &def : entity.def)
-            if (def.file_id == file_id) {
+            if (def.file_id == file_id && !Ignore(&def)) {
+              ds->kind = def.kind;
+              candidate_def_ptr = &def;
+
               if (!def.spell)
-                break;
-              ds.kind = def.kind;
+                continue;
               if (auto ls_range = GetLsRange(wfile, def.spell->extent))
-                ds.range = *ls_range;
+                ds->range = *ls_range;
               else
-                break;
+                continue;
               if (auto ls_range = GetLsRange(wfile, def.spell->range))
-                ds.selectionRange = *ls_range;
+                ds->selectionRange = *ls_range;
               else
-                break;
-              kv.first = static_cast<const void *>(&def);
+                continue;
+              def_ptr = &def;
+              break;
             }
+          
+          // Try to find a declaration.
+          if (!def_ptr && candidate_def_ptr)
```

Comment:
Forward declarations have `entity.declarations` but no `entity.def.spell`. I think they are usually unwanted, shall we apply this heuristic only for `lsSymbolKind::Namespace`?